components/server/src/workspace/gitpod-server-impl.ts

```diff
@@ -626,7 +626,15 @@ export class GitpodServerImpl implements GitpodServerWithTracing, Disposable {
 
         this.checkUser("getWorkspace");
 
-        const workspace = await this.internalGetWorkspace(workspaceId, this.workspaceDb.trace(ctx));
+        const workspace = await this.workspaceDb.trace(ctx).transaction(async (db) => {
```

Comment:
Should not we open a transaction and do call to DB only if `ideCredentialsToken` is undefined?

Reply:
We're executing this action before we know the user actually has access to the workspace. That shouldn't be the case. See line 640

Reply:
@easyCZ It is for backward compatibility with already existing workspaces. We will need to encrypt them as well and they won't go via creation anymore.

Reply:
Could you elaborate? Why would we do this without asserting the user actually owns the workspace?

We should perform as little as possible before we can assert authorization.

Reply:
Ou, you mean this method can be called by other users and then they can gain access to the secret which we use to encrypt?

Reply:
cc @mustard-mh I think @geropl brings the same point here: https://github.com/gitpod-io/gitpod/pull/16185#issuecomment-1415567896

Reply:
What I mean is that we should not create/update any information on a Workspace, which I may not be permitted to access.

In the current proposed implementation, I can make a request for ANY workspace on all of Gitpod, and it would result in the transaction execution. This is not only unathorized access, but it's actually a DDOS attack vector because the request causes asymmetrically more work for us - transaction.

Reply:
Could you guide how we could deal with it differently please? 🙏  Should we check that a user can access to this information first, check if it is not present without transaction and only then run a transaction?

Reply:
> Should not we open a transaction and do call to DB only if `ideCredentialsToken` is undefined?

@akosyakov  We maybe need also `SELECT ... FOR UPDATE` to lock the row avoid race. But I copied this update from here which has no `FOR UPDATE` https://github.com/gitpod-io/gitpod/blob/ccd64ef4779ae77512b5b8deae763af3e943ddb0/components/server/src/workspace/gitpod-server-impl.ts#L829-L846